### PR TITLE
Fix #4305 - Unstable cheats gets its own permission

### LINKED ISSUE #4305: New permission - Usage of unstable cheats
Right now, you can use the normal cheats that doesn't break the game and unstable cheats that can break the game in an instance
These should be seperated

Why:
Many people use dangerous cheats which crashes the game while still needing them for stuff like money etc

### PR DESCRIPTION
**Pull request by Nubbie**
Fix #4305
As they are dangerous for servers where cheats are needed but not the unstable that could crash yhe server, all the cheats that gives out a warning should be included here

Thanks a bunch to @Gymnasiast for squashing the commits :tada:
Thanks delebota for the support :)

### CODE REVIEW

**Comment by Nubbie** [comment]
Well if someone knows how to undo this mess and squash all these .. commits into one single one in Git Shell, that would be good if you could tell me xox

**Comment by delebota** [comment]
You'll want to follow something like this: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html
You'll want to reword the first commit and squash most of the commits into it (A few shouldn't be squashed, for example adding the train length cheat). If you don't feel confident doing that add me to your repo and I can do it for you.

**Comment by Nubbie** [comment]
@delebota That would be wonderful as I'm really stressed out now with a ton of other things I have to do before going away for a while
Im really thankful if you could sort out this mess! :smiley: 
Sending you perms

@PFCKrutonium If this gets merged, could you contact someone on Whisky to only let people who don't activate these harmful cheats to get these perms (as 90% of the crashes seems to be unstable cheats being mistreated)

File: src/network/NetworkAction.cpp [hunk 1]
```diff
@@ -218,7 +218,12 @@ const std::vector<NetworkAction> NetworkActions::Actions =
         {
             -3
         }
-    },
+    }, {
+		STR_ACTION_UNSTABLE_CHEAT, "PERMISSION_UNSTABLE_CHEATS",
+	{
+			GAME_COMMAND_UNSTABLE_CHEAT
+		}
+	},
```
Comment (Gymnasiast) [inline src/network/NetworkAction.cpp]: This block has wonky alignment.

**Comment by Gymnasiast** [comment]
Haven't tested it, but code-wise it looks good, apart from some alignment - I left a line note for that.

**Comment by Nubbie** [comment]
Noticed and tried to fix it but it got worse as usual.. sadly enough dont have time to correct that atm ;(

**Comment by Gymnasiast** [comment]
Don't worry, I have fixed it.

**Comment by Gymnasiast** [comment]
I squashed the commits into a nice big shiny strong one. ;-)

**Comment by Nubbie** [comment]
Yeah I sure hope it's a strong independent commit which doesn't require anything more from me for some days :(

File: src/network/network.h [hunk 2]
```diff
@@ -55,7 +55,7 @@ extern "C" {
 // This define specifies which version of network stream current build uses.
 // It is used for making sure only compatible builds get connected, even within
 // single OpenRCT2 version.
-#define NETWORK_STREAM_VERSION "12"
+#define NETWORK_STREAM_VERSION "13"
```
Comment (janisozaur) [inline src/network/network.h]: Requires an update now

File: src/cheats.c [hunk 3]
```diff
@@ -427,6 +423,23 @@ void game_command_cheat(int* eax, int* ebx, int* ecx, int* edx, int* esi, int* e
 	*ebx = 0;
 }
 
+void game_command_unstable_cheat(int* eax, int* ebx, int* ecx, int* edx, int* esi, int* edi, int* ebp)
+{
+	int cheat = *ecx;
```
Comment (janisozaur) [inline src/cheats.c]: Should be inside of the `if` block

**Comment by janisozaur** [comment]
This is a very coarse solution to the problem at hand. It would only let select people *toggle* the cheats, but also let everyone else *use* the cheats when enabled. While this may not be a big problem, a more fine-grained solution would check for permission when executing potentially unstable actions.

That way a cheat could be enabled server-wide, but only those with proper permissions could actually make use of it.

**Comment by Nubbie** [comment]
Never thought about it but that sounds about right, will see what could be done and maybe re-work the cheats Window a bit to hide/gray out + categorize the cheats in stability

**Comment by Nubbie** [comment]
Fucked up as usual

File: src/cheats.c [hunk 4]
```diff
@@ -427,6 +423,23 @@ void game_command_cheat(int* eax, int* ebx, int* ecx, int* edx, int* esi, int* e
...
+	if (*ebx & GAME_COMMAND_FLAG_APPLY) {
```
Comment (duncanspumpkin) [inline src/cheats.c]: It is much simpler to exit early on the not of the flag.

Comment (IntelOrca) [inline src/cheats.c]: Depends if you are following the single entry, single exit rule or not.

File: src/cheats.c [hunk 5]
```diff
@@ -417,7 +414,6 @@ void game_command_cheat(int* eax, int* ebx, int* ecx, int* edx, int* esi, int* e
 			case CHEAT_HAVEFUN: gScenarioObjectiveType = OBJECTIVE_HAVE_FUN; break;
 			case CHEAT_SETFORCEDPARKRATING: if(*edx > -1) { park_rating_spinner_value = *edx; } set_forced_park_rating(*edx); break;
 			case CHEAT_RESETDATE: date_reset(); window_invalidate_by_class(WC_BOTTOM_TOOLBAR); break;
-			case CHEAT_ALLOW_ARBITRARY_RIDE_TYPE_CHANGES: gCheatsAllowArbitraryRideTypeChanges = !gCheatsAllowArbitraryRideTypeChanges; window_invalidate_by_class(WC_RIDE); break;
```
Comment (marijnvdwerf) [inline src/cheats.c]: Is this still considered unstable since #4423 was merged? OpenRCT2 shouldn't crash anymore, but the original game will.

Comment (janisozaur) [inline src/cheats.c]: It doesn't affect only just drawing, but logic as well. In other words: yes, this is still unstable, but a bit less so.

**Comment by Nubbie** [comment]
The 'update cheats.c' commit re-worked the layout in the cheat tab rides
hey that looks pretty good

**Comment by Nubbie** [comment]
you dont know how much time i spent on reworking and rebasing here on github just to get this working now i will just have to resync that will take ages
Changed the menu a little bit

File: src/cheats.c (hunk 3 above)
Comment (Nubbie) [inline src/cheats.c]: The whole block is to look the same as 'void game_command_cheat' (line 290) which makes me wonder if both should be changed?

File: src/cheats.c (hunk 4 above)
Comment (Nubbie) [inline src/cheats.c]: The whole block is to look the same as 'void game_command_cheat' (line 290) which makes me wonder if both should be changed?

**Comment by Nubbie** [comment]
cant figure it out...

**Comment by janisozaur** [comment]
When these options were first introduced they surfaced a lot of ~~bugs~~ edge cases we didn't account for. Since then we fixed a lot, perhaps most, of them and I wouldn't consider them unstable anymore – at least the don't crash servers anymore.

Even in cases where they do crash, we'd rather fix the problem and with breakpad integrated we have good tools to inspect, so I'm going to close this.

@Nubbie if you feel this is still needed and want to contribute fixes, just open a new PR.

**Pull request closed without merging**